### PR TITLE
Fix #3486 "Order cancellation fee error"

### LINKED ISSUE #3486: Order cancellation fee error
**Describe the bug**

Why the fee to be paid is coming up in "bullion" and not BTS?

**To Reproduce**
"Just cancelling my orders and re-doing them because their about to expire, It's not doing it to all of them but more than just 1"

**Expected behavior**
Fee in BTS.

**Screenshots**
image

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**
"bug looks very random, had to cancel a lot of orders, before the fee asset switched"

### PR DESCRIPTION
**Pull request by xiangxn**
<h2>Handle issues mentioned in #3486</h2>
Closes #3486

### CODE REVIEW

File: app/lib/common/account_utils.js [hunk 1]
```diff
@@ -103,6 +114,96 @@ export default class AccountUtils {
...
+                }
+            }
+            if (hasBalance) {
+                assets.push(assetID);
+            }
+        });
+        return assets;
+    }
+
+    static getFinalFeeAsset2(
```
Comment (sschiessl-bcp) [inline app/lib/common/account_utils.js:180]: What's the difference to getFinalFeeAsset?

**Review by sschiessl-bcp — CHANGES_REQUESTED** [review]
I have the feeling that getPossibleFees2 and getFinalFeeAsset2 are really almost the same. What is the actual difference? It seems to create lots of duplicate code

File: app/lib/common/account_utils.js (hunk 1 above)
Comment (xiangxn) [reply app/lib/common/account_utils.js:180]: Although they look similar, the parameters passed in are different, and the functions called internally are also different.

Comment (sschiessl-bcp) [reply app/lib/common/account_utils.js:180]: I rebased your branch as it was not compilable before. Did you dry-code this one?

**Comment by sschiessl-bcp** [comment]
@xiangxn I have introduced a parameter to getFinalFeeAsset that allows us to not create duplicate code, please have a look at it.
Generally, your getPossibleFees was declared a getter method, but did manipulate the array or balance that the user fed into it. This should be avoided to not confuse the caller

**Comment by sschiessl-bcp** [comment]
@xiangxn please test my change again and if it still does what you wanted it to do, let's merge

**Comment by xiangxn** [comment]
> @xiangxn I have introduced a parameter to getFinalFeeAsset that allows us to not create duplicate code, please have a look at it. Generally, your getPossibleFees was declared a getter method, but did manipulate the array or balance that the user fed into it. This should be avoided to not confuse the caller

A reference parameter is required here, which will simplify the operation logic, so it is also an independent method to distinguish it from `getPossibleFees`

File: app/lib/common/account_utils.js (hunk 1 above)
Comment (xiangxn) [reply app/lib/common/account_utils.js]: Can't compile because of a PR merge

**Comment by abitmore** [comment]
>a front-end accumulated fee is required

As I understood, for a transaction which contains multiple operations, we need to sum up fees of all the operations and check if the account balance and the fee pools are sufficient to pay all the fees.

**Comment by xiangxn** [comment]
This line of code also causes subscript overflow: https://github.com/xiangxn/bitshares-ui/blob/fix-3486/app/lib/common/account_utils.js#L136

@sschiessl-bcp

**Comment by xiangxn** [comment]
I reverted the branch and removed the logic of quoting parameters, and also dealt with the problem of not compiling.

@sschiessl-bcp @abitmore

**Comment by xiangxn** [comment]
`getFinalFeeAsset2` is not compatible with `getFinalFeeAsset`, because one is for batch processing and the other is for single fee checking. If you want to force one, you may only leave `getFinalFeeAsset2`, and modify a lot of code used by the `getFinalFeeAsset` version.

My English is terrible, don't know if you can understand what I'm saying?😄

**Comment by xiangxn** [comment]
`getFinalFeeAsset2` is not compatible with `getFinalFeeAsset`, because one is for batch processing and the other is for single fee checking. If you want to force one, you may only leave `getFinalFeeAsset2`, and modify a lot of code used by the `getFinalFeeAsset` version.

My English is terrible, don't know if you can understand what I'm saying?😄

**Comment by sschiessl-bcp** [comment]
I don't exactly understand yet why we need this complicated logic and duplicate code. 

I have restored my change to a new branch so allow comparison, I have also fixed the subscript overflow. 
https://github.com/bitshares/bitshares-ui/tree/bugfix/3486_cancel_default_fee_selection

I probably don't understand it correctly at the moment, can you please tell me
 a) what behavior it is that you want to have
 b) and in which case my code does not deliver it?
 
 In any case, we need to avoid duplicate code if it is at the end just a very small change to the business logic.

**Comment by xiangxn** [comment]
In this problem, we need to take out the user's balance at one time and accumulate the required fees at the front end. The original method is to take the data on the chain every time, so it cannot achieve the purpose.

**Comment by sschiessl-bcp** [comment]
Ah, now I understand. My way was doing the check only for one op, but we want it for ALL ops in the loop.

**Pull request merged**